File: python/123.py
"""Project Euler Problem 123.

Let p_n be the nth prime: 2, 3, 5, 7, 11, ..., and let r be the remainder when
(p_n - 1)^n + (p_n + 1)^n is divided by p_n^2.

For n even, remainder is 2.
For n odd, remainder is 2*n*p_n mod p_n^2.
For large enough n, 2*n < p_n, so remainder is simply 2*n*p_n.

We want the smallest n such that remainder > 10^10.
"""

def primes_sieve(limit):
    """Return a list of primes up to limit using Sieve of Eratosthenes."""
    is_prime = bytearray(b'\x01') * (limit + 1)
    is_prime[0] = is_prime[1] = 0
    for i in range(2, int(limit**0.5) + 1):
        if is_prime[i]:
            is_prime[i*i::i] = bytearray(len(is_prime[i*i::i]))
    return [i for i in range(2, limit + 1) if is_prime[i]]

def solve():
    target = 10**10
    # We need roughly the first 25000 primes. The 25000th prime is around
    # 287000, so a sieve up to 300000 is more than enough.
    primes = primes_sieve(300000)

    for n in range(1, len(primes) + 1, 2):  # Only check odd n
        p_n = primes[n - 1]  # 1-indexed: p_1 = 2, p_2 = 3, ...
        remainder = 2 * n * p_n
        if remainder > target:
            return n

    return None

if __name__ == "__main__":
    print(solve())

File: c/123.c
/* Project Euler Problem 123: Prime square remainders.
 *
 * For odd n: remainder = 2*n*p_n mod p_n^2 = 2*n*p_n (when 2n < p_n).
 * Find smallest odd n such that 2*n*p_n > 10^10.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

#define SIEVE_LIMIT 300000

int main(void) {
    /* Sieve of Eratosthenes */
    bool *is_prime = calloc(SIEVE_LIMIT + 1, sizeof(bool));
    if (!is_prime) return 1;

    for (int i = 2; i <= SIEVE_LIMIT; i++) is_prime[i] = true;
    for (int i = 2; (long long)i * i <= SIEVE_LIMIT; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= SIEVE_LIMIT; j += i)
                is_prime[j] = false;
        }
    }

    /* Collect primes */
    int *primes = malloc(30000 * sizeof(int));
    int nprimes = 0;
    for (int i = 2; i <= SIEVE_LIMIT; i++) {
        if (is_prime[i]) primes[nprimes++] = i;
    }

    long long target = 10000000000LL; /* 10^10 */

    for (int n = 1; n <= nprimes; n += 2) { /* odd n only */
        long long p_n = primes[n - 1]; /* 1-indexed */
        long long remainder = 2LL * n * p_n;
        if (remainder > target) {
            printf("%d\n", n);
            free(is_prime);
            free(primes);
            return 0;
        }
    }

    free(is_prime);
    free(primes);
    return 1;
}
